crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+    pub fn algorithm(&self) -> &Algorithm {
+        &self.algorithm
+    }
+
+    /// Emit TSIG RR and RDATA as used for computing MAC
+    ///
+    /// ```text
+    /// 4.3.3.  TSIG Variables
+    ///
+    ///    Also included in the digest is certain information present in the
+    ///    TSIG RR.  Adding this data provides further protection against an
+    ///    attempt to interfere with the message.
+    ///
+    ///    +============+================+====================================+
+    ///    | Source     | Field Name     | Notes                              |
+    ///    +============+================+====================================+
+    ///    | TSIG RR    | NAME           | Key name, in canonical wire format |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RR    | CLASS          | MUST be ANY                        |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RR    | TTL            | MUST be 0                          |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Algorithm Name | in canonical wire format           |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Time Signed    | in network byte order              |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Fudge          | in network byte order              |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Error          | in network byte order              |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Other Len      | in network byte order              |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Other Data     | exactly as transmitted             |
+    ///    +------------+----------------+------------------------------------+
+    /// ```
+    pub fn emit_tsig_for_mac(
+        &self,
+        encoder: &mut BinEncoder<'_>,
+        key_name: &Name,
+    ) -> ProtoResult<()> {
+        key_name.emit_as_canonical(encoder, true)?;
+        DNSClass::ANY.emit(encoder)?;
+        encoder.emit_u32(0)?; // TTL
+        self.algorithm.emit(encoder)?;
+        encoder.emit_u16((self.time >> 32) as u16)?;
+        encoder.emit_u32(self.time as u32)?;
+        encoder.emit_u16(self.fudge)?;
+        encoder.emit_u16(self.error)?;
+        encoder.emit_u16(self.other.len() as u16)?;
+        encoder.emit_vec(&self.other)?;
+        Ok(())
+    }
+
+    /// Add actual MAC value to existing TSIG record data.
+    ///
+    /// # Arguments
+    ///
+    /// * `mac` - mac to be stored in this record.
+    pub fn set_mac(self, mac: Vec<u8>) -> Self {
+        assert!(mac.len() < (1 << 16));
```

Comment:
same comment as above in regards to the potential for this being abused during deserialization.

Reply:
it's not used in deserialization either, and it's not an issue for the same reason as above

Reply:
Similar comment, please add the Panic section to the API doc in that case